### PR TITLE
Basic HLSL compute shader support

### LINKED ISSUE #249: Implement SSBO flattening for HLSL
We will need to use ByteAddressBuffer and RWAddressBuffer to implement SSBOs. This will require special handling of SSBO access to handle address flattening not only for loads, but also for stores.

### PR DESCRIPTION
**Pull request by HansKristian-ARM**
Built upon https://github.com/KhronosGroup/SPIRV-Cross/pull/220.
Fixes #249.

Lots of stuff is certainly missing, but it should be usable now for basic compute shaders.

### CODE REVIEW

File: spirv_hlsl.cpp [hunk 1]
```diff
@@ -1798,6 +1850,199 @@ void CompilerHLSL::emit_glsl_op(uint32_t result_type, uint32_t id, uint32_t eop,
...
+		load_op = "Load3";
+		break;
+	case 4:
+		load_op = "Load4";
+		break;
+	default:
+		SPIRV_CROSS_THROW("Unknown vector size.");
+	}
+
+	auto load_expr = join(chain.base, ".", load_op, "(", chain.dynamic_index, chain.static_index, ")");
```
Comment (Kangz) [inline spirv_hlsl.cpp]: Is a + missing between dynamic_index and static_index here? The same happens in the handling of store.

Comment (HansKristian-ARM) [inline spirv_hlsl.cpp]: This is intentional. dynamic_index is something like "16 * foo + ", or blank.

Comment (Kangz) [inline spirv_hlsl.cpp]: Sounds good

**Comment by Kangz** [comment]
Thanks for taking a look at this, approach looks like the proper way to support SSBOs. How will passing SPIRCAccessChains to functions work, will it pass a pair of (buffer, offset)?

**Comment by HansKristian-ARM** [comment]
I'm not entirely sure if you can pass access chains like that to functions, but glslang codegen typically goes via a dummy variable to satisfy the wonky requirements of function calls in GLSL/HLSL. Either we do that as well if needed, or pass down Buffer + Offset. The annoying bit with latter of course is that you probably cannot cast between ByteAddressBuffer and RWByteAddressBuffer, hnnngggggg. <_<

**Comment by Kangz** [comment]
Pulling @dneto0 into the conversation. From what I remember, there is a constraint of SPIRV that makes it so we can know in the function if the parameter is a pointer in a readable or RW SSBO, but I can't find it in the spec. Do you know about it?

**Pull request merged**